usersession/userd/launcher_internal_test.go

```diff
@@ -0,0 +1,120 @@
+	"/var/lib/snapd/desktop/applications/multipass_multipass-gui.desktop",
+	"/var/lib/snapd/desktop/applications/cevelop_cevelop.desktop",
+	"/var/lib/snapd/desktop/applications/egmde-confined-desktop_egmde-confined-desktop.desktop",
+	"/var/lib/snapd/desktop/applications/classic-snap-analyzer_classic-snap-analyzer.desktop",
+	"/var/lib/snapd/desktop/applications/vlc_vlc.desktop",
+	"/var/lib/snapd/desktop/applications/gnome-calculator_gnome-calculator.desktop",
+	"/var/lib/snapd/desktop/applications/mir-kiosk-kodi_mir-kiosk-kodi.desktop",
+	"/var/lib/snapd/desktop/applications/gnome-characters_gnome-characters.desktop",
+	"/var/lib/snapd/desktop/applications/clion_clion.desktop",
+	"/var/lib/snapd/desktop/applications/gnome-system-monitor_gnome-system-monitor.desktop",
+	"/var/lib/snapd/desktop/applications/inkscape_inkscape.desktop",
+	"/var/lib/snapd/desktop/applications/gnome-logs_gnome-logs.desktop",
+
+	"/var/lib/snapd/desktop/applications/foo-bar/baz.desktop",
+	"/var/lib/snapd/desktop/applications/baz/foo-bar.desktop",
+}
+
+func existsOnMockFileSystem(desktop_file string) bool {
+	return strutil.ListContains(mockFileSystem, desktop_file)
+}
+
+func (s *launcherInternalSuite) TestDesktopFileIDToFilenameSucceedsWithValidId(c *C) {
+
+	var desktopIdTests = []struct {
+		id     string
+		expect string
+	}{
+		{"mir-kiosk-scummvm_mir-kiosk-scummvm.desktop", "/var/lib/snapd/desktop/applications/mir-kiosk-scummvm_mir-kiosk-scummvm.desktop"},
+		{"foo-bar-baz.desktop", "/var/lib/snapd/desktop/applications/foo-bar/baz.desktop"},
+		{"baz-foo-bar.desktop", "/var/lib/snapd/desktop/applications/baz/foo-bar.desktop"},
+	}
+
+	for _, test := range desktopIdTests {
+		actual, err := desktopFileIDToFilename(existsOnMockFileSystem, test.id)
+		c.Assert(err, IsNil)
+		c.Assert(actual, Equals, test.expect)
+	}
+}
+
+func (s *launcherInternalSuite) TestDesktopFileIDToFilenameFailsWithInvalidId(c *C) {
+	var desktopIdTests = []string{
+		"mir-kiosk-scummvm-mir-kiosk-scummvm.desktop",
+		"bar-foo-baz.desktop",
+		"bar-baz-foo.desktop",
+		"foo-bar_foo-bar.desktop",
+	}
+
+	for _, id := range desktopIdTests {
+		_, err := desktopFileIDToFilename(existsOnMockFileSystem, id)
+		c.Assert(err, NotNil)
+	}
+}
+
+func (s *launcherInternalSuite) TestParseExecCommandSucceedsWithValidEntry(c *C) {
+	var exec_command = []struct {
+		exec_command string
+		expect       []string
+	}{
+		{"env BAMF_DESKTOP_FILE_HINT=/var/lib/snapd/desktop/applications/mir-kiosk-scummvm_mir-kiosk-scummvm.desktop /snap/bin/mir-kiosk-scummvm %U",
+			[]string{"env", "BAMF_DESKTOP_FILE_HINT=/var/lib/snapd/desktop/applications/mir-kiosk-scummvm_mir-kiosk-scummvm.desktop", "/snap/bin/mir-kiosk-scummvm"}},
+		{"/snap/bin/foo -f %U %%bar", []string{"/snap/bin/foo", "-f", "%bar"}},
+		{"/snap/bin/foo '-f %U %%bar'", []string{"/snap/bin/foo", "-f %U %%bar"}},
+		{"/snap/bin/foo \"'-f bar'\"", []string{"/snap/bin/foo", "'-f bar'"}},
```

Comment:
I misread this test so I think it might help to group things with comments. I suggest (which adds some tests):
```
                // valid with no exec variables
                {"env BAMF_DESKTOP_FILE_HINT=/var/lib/snapd/desktop/applications/mir-kiosk-scummvm_mir-kiosk-scummvm.desktop /snap/bin/mir-kiosk-scummvm %U",
                        []string{"env", "BAMF_DESKTOP_FILE_HINT=/var/lib/snapd/desktop/applications/mir-kiosk-scummvm_mir-kiosk-scummvm.desktop", "/snap/bin/mir-kiosk-scummvm"}},
                // valid with literal '%' and no exec variables
                {"/snap/bin/foo -f %%bar", []string{"/snap/bin/foo", "-f", "%bar"}},
                {"/snap/bin/foo -f %%bar %%baz", []string{"/snap/bin/foo", "-f", "%bar", "%baz"}},
                // valid with quoted strings are passed through
                {"/snap/bin/foo '-f %U'", []string{"/snap/bin/foo", "-f %U"}},
                {"/snap/bin/foo '-f %%bar'", []string{"/snap/bin/foo", "-f %%bar"}},
                {"/snap/bin/foo '-f %U %%bar'", []string{"/snap/bin/foo", "-f %U %%bar"}},
                {"/snap/bin/foo \"'-f bar'\"", []string{"/snap/bin/foo", "'-f bar'"}},
                {"/snap/bin/foo '\"-f bar\"'", []string{"/snap/bin/foo", "\"-f bar\""}},
                // valid with exec variables stripped out
                {"/snap/bin/foo -f %U", []string{"/snap/bin/foo", "-f"}},
                {"/snap/bin/foo -f %U %Y", []string{"/snap/bin/foo", "-f"}},
                {"/snap/bin/foo -f %U bar", []string{"/snap/bin/foo", "-f", "bar"}},
                {"/snap/bin/foo -f %U bar %Y", []string{"/snap/bin/foo", "-f", "bar"}},
                // valid with mixture of literal '%' and exec variables
                {"/snap/bin/foo -f %U %%bar", []string{"/snap/bin/foo", "-f", "%bar"}},
                {"/snap/bin/foo -f %U %Y %%bar", []string{"/snap/bin/foo", "-f", "%bar"}},
                {"/snap/bin/foo -f %U %%bar %Yr", []string{"/snap/bin/foo", "-f", "%bar"}},
                {"/snap/bin/foo -f %%bar %U %Y", []string{"/snap/bin/foo", "-f", "%bar"}},
        }
```